aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/operation/AWSS3StoragePathDownloadFileOperation.kt

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amplifyframework.storage.s3.operation
+
+import com.amplifyframework.auth.AuthCredentialsProvider
+import com.amplifyframework.core.Amplify
+import com.amplifyframework.core.Consumer
+import com.amplifyframework.hub.HubChannel
+import com.amplifyframework.hub.HubEvent
+import com.amplifyframework.storage.IdentityIdProvidedStoragePath
+import com.amplifyframework.storage.StorageChannelEventName
+import com.amplifyframework.storage.StorageException
+import com.amplifyframework.storage.StoragePathValidationException
+import com.amplifyframework.storage.StringStoragePath
+import com.amplifyframework.storage.TransferState
+import com.amplifyframework.storage.operation.StorageDownloadFileOperation
+import com.amplifyframework.storage.result.StorageDownloadFileResult
+import com.amplifyframework.storage.result.StorageTransferProgress
+import com.amplifyframework.storage.s3.extensions.invalidStoragePathException
+import com.amplifyframework.storage.s3.extensions.unsupportedStoragePathException
+import com.amplifyframework.storage.s3.request.AWSS3StoragePathDownloadFileRequest
+import com.amplifyframework.storage.s3.service.StorageService
+import com.amplifyframework.storage.s3.transfer.TransferListener
+import com.amplifyframework.storage.s3.transfer.TransferObserver
+import java.io.File
+import java.util.UUID
+import java.util.concurrent.ExecutorService
+import kotlinx.coroutines.runBlocking
+
+/**
+ * An operation to download a file from AWS S3.
+ */
+internal class AWSS3StoragePathDownloadFileOperation(
+    private val transferId: String = UUID.randomUUID().toString(),
+    private val request: AWSS3StoragePathDownloadFileRequest?,
```

Comment:
How can request be null?

Reply:
This will be legitimate. I'll need to make sure other new operations also have nullable requests as these are constructed from the `getTransfer` api where we no longer have the request.

Reply:
The reason for this is because the getTransfer API cannot reconstruct the request.

However, after discussions, we have determined that in case a customer as explicitly casting to the AWS operation type, we will not adjust the operations we create in getTransfer. 

We can do this, because there is no difference in behavior between the old operations transfer and the new transfer operations that use StoragePath, once the transfer has been initiated.

We cannot distinguish whether the old or new (StoragePath) operations were used to start the request when reconstructing the operation from the local db.

Reply:
As a result to all of this, the new operations will have non-null requests for now, as we do not need them to be nullable.